### PR TITLE
Add `SecondMinMax`

### PR DESCRIPTION
**Pull request by BharathSanjeeviT**
### Description 
This PR provides a feature that contains the Java Program to find second largest / smallest element from the array with the most optimal solution

#### This PR addresses the Issue #4430 

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.

### CODE REVIEW

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package com.thealgorithms.maths;
+
+public class SecondMinMax {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:3]: You can make a _proper utility class_ of this.

```suggestion
public final class SecondMinMax {
    private SecondMinMax() {
    }
```

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    /**
+     * @brief Finds the Second minimum / maximum value from the array
+     * @param int array
+     * @exception IllegalArgumentException if input array is empty or length 1 also if all elements are same
+     * @return the second minimum / maximum value from the input array
+     * @author Bharath Sanjeevi ( https://github.com/BharathSanjeeviT )
+     */
+
+    public static int findSecondMin(int[] arr) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:13]:
```suggestion
    public static int findSecondMin(final int[] arr) {
```

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                if (num < min) {
+                    secMin = min;
+                    min = num;
+                } else if ((num < secMin) && (num != min))
+                    secMin = num;
+        if (secMin == Integer.MAX_VALUE) throw new IllegalArgumentException("Cant find Second Minimum in array full of same numbers");
+        return secMin;
+    }
+
+    public static int findSecondMax(int[] arr) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:30]:
```suggestion
    public static int findSecondMax(final int[] arr) {
```

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        int secMax = Integer.MIN_VALUE, max = Integer.MIN_VALUE;
+        if (arr.length == 0)
+            throw new IllegalArgumentException("Cant find Second Maximum for empty array");
+        else if (arr.length == 1)
+            throw new IllegalArgumentException("Cant find Second Maximum for single element");
+        else
+            for (int num : arr)
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:37]:
```suggestion
            for (final int num : arr)
```

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+     */
+
+    public static int findSecondMin(int[] arr) {
+        int secMin = Integer.MAX_VALUE, min = Integer.MAX_VALUE;
+        if (arr.length == 0)
+            throw new IllegalArgumentException("Cant find Second Minimum for empty array");
+        else if (arr.length == 1)
+            throw new IllegalArgumentException("Cant find Second Minimum for single element");
+        else
+            for (int num : arr)
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:20]:
```suggestion
            for (final int num : arr)
```

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+     * @return the second minimum / maximum value from the input array
+     * @author Bharath Sanjeevi ( https://github.com/BharathSanjeeviT )
+     */
+
+    public static int findSecondMin(int[] arr) {
+        int secMin = Integer.MAX_VALUE, min = Integer.MAX_VALUE;
+        if (arr.length == 0)
+            throw new IllegalArgumentException("Cant find Second Minimum for empty array");
+        else if (arr.length == 1)
+            throw new IllegalArgumentException("Cant find Second Minimum for single element");
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:18]: This logic is repeated in both methods - maybe you can abstract it to some private method like `checkInput(...)`.

**Review by vil02 — CHANGES_REQUESTED** [review]
@BharathSanjeeviT thanks for an interesting contribution. Before we will marge it, I have some remarks.

Please add some proper tests. Good examples to look at:
- [`Factorial`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/main/java/com/thealgorithms/maths/Factorial.java) and [`FactorialTest`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/test/java/com/thealgorithms/maths/FactorialTest.java),
- [`FindMax`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/main/java/com/thealgorithms/maths/FindMax.java) and [`FindMaxTest`](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/test/java/com/thealgorithms/maths/FindMaxTest.java).

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 1 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java:3]: Sure will change this 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 2 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java:13]: Sure will change this 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 3 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java:30]: Sure will change this 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 6 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java:18]: Sure will change this 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 5 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java:20]: Sure will change this 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 4 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java:37]: Sure will change this 👍

**Comment by BharathSanjeeviT** [comment]
Hey @vil02 , Thank you for looking into my contribution and suggesting all the changes. Hope I did correct all those you said 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+                secMax = num;
+        if (secMax == Integer.MIN_VALUE) throw new IllegalArgumentException("Cant find Second Maximum / Minimum in array full of same numbers");
+        return secMax;
+    }
+
+    private static void checkInput(final int[] arr) {
+        if (arr.length == 0)
+            throw new IllegalArgumentException("Cant find Second Maximum / Minimum for empty array");
+        else if (arr.length == 1)
+            throw new IllegalArgumentException("Cant find Second Maximum / Minimum for single element");
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:46]: What do you think about the following simplification?
```suggestion
        if (arr.length < 2) {
            throw new IllegalArgumentException("Input array must have length of at least two.");
        }
```

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+                min = num;
+            } else if ((num < secMin) && (num != min))
+                secMin = num;
+        if (secMin == Integer.MAX_VALUE) throw new IllegalArgumentException("Cant find Second Maximum / Minimum in array full of same numbers");
+        return secMin;
+    }
+
+    public static int findSecondMax(final int[] arr) {
+        int secMax = Integer.MIN_VALUE, max = Integer.MIN_VALUE;
+        checkInput(arr);
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:31]:
```suggestion
        checkInput(arr);
        int secMax = Integer.MIN_VALUE, max = Integer.MIN_VALUE;
```

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+     * @return the second minimum / maximum value from the input array
+     * @author Bharath Sanjeevi ( https://github.com/BharathSanjeeviT )
+     */
+
+    private SecondMinMax() {
+    }
+
+    public static int findSecondMin(final int[] arr) {
+        int secMin = Integer.MAX_VALUE, min = Integer.MAX_VALUE;
+        checkInput(arr);
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:18]:
```suggestion
        checkInput(arr);
        int secMin = Integer.MAX_VALUE, min = Integer.MAX_VALUE;
```

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        for (final int num : arr)
+            if (num < min) {
+                secMin = min;
+                min = num;
+            } else if ((num < secMin) && (num != min))
+                secMin = num;
+        if (secMin == Integer.MAX_VALUE) throw new IllegalArgumentException("Cant find Second Maximum / Minimum in array full of same numbers");
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:25]: What do you think about introducing a method `checkOutput(...)`? I thnk it it enough to check if _best_ and _secondBest_ are the same (here `min` and `secondMin`).

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+    public static int findSecondMin(final int[] arr) {
+        int secMin = Integer.MAX_VALUE, min = Integer.MAX_VALUE;
+        checkInput(arr);
+        for (final int num : arr)
+            if (num < min) {
+                secMin = min;
+                min = num;
+            } else if ((num < secMin) && (num != min))
+                secMin = num;
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:24]:
```suggestion
            } else if ((num < secMin) && (num != min)) {
                secMin = num;
            }
```

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+    public static int findSecondMax(final int[] arr) {
+        int secMax = Integer.MIN_VALUE, max = Integer.MIN_VALUE;
+        checkInput(arr);
+        for (final int num : arr)
+            if (num > max) {
+                secMax = max;
+                max = num;
+            } else if ((num > secMax) && (num != max))
+                secMax = num;
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:37]:
```suggestion
            } else if ((num > secMax) && (num != max)) {
                secMax = num;
            }
```

**Review by BharathSanjeeviT — COMMENTED** [review]
Hope I did solved all the change that you have addressed 👍

File: src/test/java/com/thealgorithms/maths/SecondMinMaxTest.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+        assertEquals(exception.getMessage(), EXP_MSG_ARR_SAME_ELE);
+    }
+
+    @Test
+    public void testForValidOutput() {
+        assertEquals(2, SecondMinMax.findSecondMin(new int[] {1, 2}));
+        assertEquals(1, SecondMinMax.findSecondMax(new int[] {1, 2}));
+        assertEquals(2, SecondMinMax.findSecondMin(new int[] {10, 1, 2, 4, 6, 3, 5, 7, 3, 2, 5, 7, 4, 2, 1, 212, 343, 565, 2232, 65565}));
+        assertEquals(2232, SecondMinMax.findSecondMax(new int[] {10, 1, 2, 4, 6, 3, 5, 7, 3, 2, 5, 7, 4, 2, 1, 212, 343, 565, 2232, 65565}));
+    }
```
Comment (vil02) [inline src/test/java/com/thealgorithms/maths/SecondMinMaxTest.java:38]: What do you think about making some _parametrized tests_ here like in [FindMaxTest.java](https://github.com/TheAlgorithms/Java/blob/566c27a996f48a94916170c159ae735546b39c42/src/test/java/com/thealgorithms/maths/FindMaxTest.java). The input stream could return object of a class having such fields like `inputArray`, `min`, `secondMin`, `max`, `secondMax`.

**Review by vil02 — CHANGES_REQUESTED** [review]
The logic in `findSecondMin` and `findSecondMax` is very similar (basically the _type of inequality_). Do you see some way to remove duplicated code?

By the way, you can commit suggestions directly just by clicking the `Commit suggestion` button - this automatically resolves the conversation.

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 7 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java]: Well, this change looks good 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 8 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java]: will change this 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 10 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java]: Yeah, this change looks good 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 11 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java]: will change this 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 12 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java]: will change this 👍

File: src/test/java/com/thealgorithms/maths/SecondMinMaxTest.java (hunk 13 above)
Comment (BharathSanjeeviT) [reply src/test/java/com/thealgorithms/maths/SecondMinMaxTest.java]: Yeah made a method like the file you have mentioned 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 9 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java]: will change this 👍

**Comment by BharathSanjeeviT** [comment]
Hey @vil02 , made the changes that you have mentioned and btw i thought *Commit Changes* would add extra commits, thanks for the information 👍

**Review by BharathSanjeeviT — COMMENTED** [review]
Made all these changes 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+
+    private static void checkInput(final int[] arr) {
+        if (arr.length < 2) throw new IllegalArgumentException("Input array must have length of at least two");
+    }
+
+    private static int checkOutput(final int secNum) {
+        if ((secNum == Integer.MAX_VALUE) || (secNum == Integer.MIN_VALUE)) throw new IllegalArgumentException("Input array should have at least 2 distinct elements");
+        return secNum;
+    }
+}
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:50]: As suggested in [here](https://github.com/TheAlgorithms/Java/pull/4432#pullrequestreview-1651654653):

```suggestion
package com.thealgorithms.maths;

import java.util.function.BiPredicate;

public final class SecondMinMax {

    /**
     * @brief Utility class finding second smallest or gatest eleent in the input array
     * @author Bharath Sanjeevi ( https://github.com/BharathSanjeeviT )
     */

    private SecondMinMax() {
    }

    private static int secondBest(final int[] arr, final int initialVal, final BiPredicate<Integer, Integer> isBetter) {
        checkInput(arr);
        int best = initialVal;
        int secBest = initialVal;
        for (final int num : arr) {
            if (isBetter.test(num, best)) {
                secBest = best;
                best = num;
            } else if ((isBetter.test(num, secBest)) && (num != best)) {
                secBest = num;
            }
        }
        checkOutput(initialVal, secBest);
        return secBest;
    }

    /**
     * @brief Finds the Second smallest value from the array
     * @param arr the input array
     * @exception IllegalArgumentException if input array is of length less than 2 also if all elements are same
     * @return the second smallest value from the input array
     */
    public static int findSecondMin(final int[] arr) {
        return secondBest(arr, Integer.MAX_VALUE, (a, b) -> a < b);
    }

    /**
     * @brief Finds the Second largest value from the array
     * @param arr the input array
     * @exception IllegalArgumentException if input array is of length less than 2 also if all elements are same
     * @return the second largest value from the input array
     */
    public static int findSecondMax(final int[] arr) {
        return secondBest(arr, Integer.MIN_VALUE, (a, b) -> a > b);
    }

    private static void checkInput(final int[] arr) {
        if (arr.length < 2) {
            throw new IllegalArgumentException("Input array must have length of at least two");
        }
    }

    private static void checkOutput(final int initialVal, final int secBest) {
        if (secBest == initialVal) {
            throw new IllegalArgumentException("Input array should have at least 2 distinct elements");
        }
    }
}
```

File: src/test/java/com/thealgorithms/maths/SecondMinMaxTest.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    @ParameterizedTest
+    @MethodSource("inputStreamMax")
+    void numberTestsMax(int expected, int[] input) {
+        Assertions.assertEquals(expected, SecondMinMax.findSecondMax(input));
+    }
+
+    private static Stream<Arguments> inputStreamMax() {
+        return Stream.of(Arguments.of(9, new int[] {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}), Arguments.of(4, new int[] {5, 5, 4, 5, 5}), Arguments.of(-1, new int[] {-1, 0}), Arguments.of(-2, new int[] {-10, -9, -8, -7, -6, -5, -4, -3, -2, -1}), Arguments.of(8, new int[] {3, -2, 3, 9, -4, -4, 8}));
+    }
+}
```
Comment (vil02) [inline src/test/java/com/thealgorithms/maths/SecondMinMaxTest.java:55]: as suggested in [here](https://github.com/TheAlgorithms/Java/pull/4432#discussion_r1341925409):
```suggestion
package com.thealgorithms.maths;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.stream.Stream;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

public class SecondMinMaxTest {

    private static final String EXP_MSG_ARR_LEN_LESS_2 = "Input array must have length of at least two";
    private static final String EXP_MSG_ARR_SAME_ELE = "Input array should have at least 2 distinct elements";

    public static class TestCase {
        public TestCase(final int[] inInputArray, final int inSecondMin, final int inSecondMax) {
            inputArray = inInputArray;
            secondMin = inSecondMin;
            secondMax = inSecondMax;
        }
        final int[] inputArray;
        final int secondMin;
        final int secondMax;
    }

    @Test
    public void testForEmptyInputArray() {
        IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> SecondMinMax.findSecondMin(new int[] {}));
        assertEquals(exception.getMessage(), EXP_MSG_ARR_LEN_LESS_2);
    }

    @Test
    public void testForArrayWithSingleElement() {
        IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> SecondMinMax.findSecondMax(new int[] {1}));
        assertEquals(exception.getMessage(), EXP_MSG_ARR_LEN_LESS_2);
    }

    @Test
    public void testForArrayWithSameElements() {
        IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> SecondMinMax.findSecondMin(new int[] {1, 1, 1, 1}));
        assertEquals(exception.getMessage(), EXP_MSG_ARR_SAME_ELE);
    }

    @ParameterizedTest
    @MethodSource("inputStream")
    void numberTestsMin(final TestCase tc) {
        assertEquals(tc.secondMin, SecondMinMax.findSecondMin(tc.inputArray));
    }

    @ParameterizedTest
    @MethodSource("inputStream")
    void numberTestsMax(final TestCase tc) {
        assertEquals(tc.secondMax, SecondMinMax.findSecondMax(tc.inputArray));
    }

    private static Stream<Arguments> inputStream() {
        return Stream.of(Arguments.of(new TestCase(new int[] {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 2, 9)), Arguments.of(new TestCase(new int[] {5, 4, 5, 5, 5}, 5, 4)), Arguments.of(new TestCase(new int[] {-1, 0}, 0, -1)),
            Arguments.of(new TestCase(new int[] {-10, -9, -8, -7, -6, -5, -4, -3, -2, -1}, -9, -2)), Arguments.of(new TestCase(new int[] {3, -2, 3, 9, -4, -4, 8}, -2, 8)));
    }
}
```

**Review by vil02 — CHANGES_REQUESTED** [review]
Sorry for a _monster suggestion_: sometimes it is easier to express ideas just by writing code.

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 14 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java]: Sure will update this 👍

File: src/test/java/com/thealgorithms/maths/SecondMinMaxTest.java (hunk 15 above)
Comment (BharathSanjeeviT) [reply src/test/java/com/thealgorithms/maths/SecondMinMaxTest.java]: Sure will update this 👍

**Comment by BharathSanjeeviT** [comment]
Hey @vil02 , made the changes that you've suggested 👍

**Comment by BharathSanjeeviT** [comment]
> Hey @vil02 , made the changes that you've suggested 👍



> Sorry for a _monster suggestion_: sometimes it is easier to express ideas just by writing code.

The changes you say, actually making me to learn new things, so i should thank you 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    private static void checkInput(final int[] arr) {
+        if (arr.length < 2) throw new IllegalArgumentException("Input array must have length of at least two");
+    }
+
+    private static int checkOutput(final int secNum, int initialVal) {
+        if (secNum == initialVal) throw new IllegalArgumentException("Input array should have at least 2 distinct elements");
+        return secNum;
+    }
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:55]: I would be consistent and make both of the `checkSomething` methods to have return type `void`. I would also suggest to use `{` and `}` even for _single liners_ (helps avoid some misstakes).

```suggestion
    private static void checkOutput(final int secNum, final int initialVal) {
        if (secNum == initialVal) {
            throw new IllegalArgumentException("Input array should have at least 2 distinct elements");
        }
    }
```

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    public static int findSecondMin(final int[] arr) {
+        return secondBest(arr, Integer.MAX_VALUE, (a, b) -> a < b);
+    }
+
+    public static int findSecondMax(final int[] arr) {
+        return secondBest(arr, Integer.MIN_VALUE, (a, b) -> a > b);
+    }
+
+    private static void checkInput(final int[] arr) {
+        if (arr.length < 2) throw new IllegalArgumentException("Input array must have length of at least two");
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:49]:
```suggestion
        if (arr.length < 2) {
            throw new IllegalArgumentException("Input array must have length of at least two");
        }
```

**Review by vil02 — CHANGES_REQUESTED** [review]
It is really close!

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 16 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java]: Changed @vil02 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java (hunk 17 above)
Comment (BharathSanjeeviT) [reply src/main/java/com/thealgorithms/maths/SecondMinMax.java]: Changes @vil02 👍

**Comment by BharathSanjeeviT** [comment]
Hey @vil02 , the changes you have suggested are made, hope you merge this PR 👍

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+        return secondBest(arr, Integer.MIN_VALUE, (a, b) -> a > b);
+    }
+
+    private static void checkInput(final int[] arr) {
+        if (arr.length < 2) {
+            throw new IllegalArgumentException("Input array must have length of at least two");
+        }
+    }
+
+    private static void checkOutput(final int secNum, int initialVal) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:54]:
```suggestion
    private static void checkOutput(final int secNum, final int initialVal) {
```

File: src/main/java/com/thealgorithms/maths/SecondMinMax.java [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+     * @return the second minimum / maximum value from the input array
+     * @author Bharath Sanjeevi ( https://github.com/BharathSanjeeviT )
+     */
+
+    private SecondMinMax() {
+    }
+
+    private static int secondBest(final int[] arr, final int initialVal, final BiPredicate<Integer, Integer> isBetter) {
+        checkInput(arr);
+        int best = initialVal, secBest = initialVal;
```
Comment (vil02) [inline src/main/java/com/thealgorithms/maths/SecondMinMax.java:19]:
```suggestion
        int best = initialVal;
        int secBest = initialVal;
```

**Review by vil02 — APPROVED** [review]

**Comment by vil02** [comment]
@BharathSanjeeviT thanks for your contribution!

**Comment by BharathSanjeeviT** [comment]
Hey @vil02 , Thank you soo much, i loved this process of contributing towards The Algorithms 👍

**Pull request merged**